### ISSUE TITLE
Add a public API like twisted.internet.test.connectionmixins.serverFactoryFor

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5016]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#5016|
|Type|enhancement|
|Created|2011-04-03 16:52:12Z|
|Branch|https://github.com/twisted/twisted/tree/factory-for-5016|

These two lines are simple but typing them out gets tedious:

```python
    factory = ServerFactory()
    factory.protocol = protocol
```

It would be nice to have this API provided by Twisted somewhere.  Ideally it would just be `ServerFactory(protocol)`, but `ServerFactory` has no `__init__` and giving it one breaks this pattern:

```python
class Foo:
    def __init__(self):
        self.stuff = things

class Bar(ServerFactory, Foo):
    pass
```

(since suddenly `Bar.__init__` means `ServerFactory.__init__` instead of `Foo.__init__`).

Perhaps we could add a class method, though.  `ServerFactory.forProtocol(protocol, *args, **kwargs)`.

Maybe this belongs on `Factory` rather than `ServerFactory`, too.

Attachments:

* [forProtocol.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/c5f/c5f9eb3ff78f50a1061e4c9d8e3348b8b17b5eee/86165a9bc0b9537e0613000fa0ec7cd42eb7ef8e.patch) (12204 bytes) - added by Julian on 2013-03-24 14:49:02Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5016 5016
type__enhancement enhancement
reporter__exarkun exarkun
priority__low low
milestone__ 
branch__branches_factory_for_5016 branches/factory-for-5016
branch_author__therve therve
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1301849532000000 1301849532000000
changetime__1364484984000000 1364484984000000
version__None None
owner__therve therve

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_1"></a>|@Julian commented|
|-|-|

Here's a patch along with some changes in a few places that this cleans up a tiny bit.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

(In [37858]) Branching to 'factory-for-5016'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

Looks good, applying!

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set status to `closed`|
|-|-|

(In [37864]) Merge factory-for-5016

Author: Julian
Reviewer: therve
Fixes: [#5016](https://github.com/twisted/twisted/issues/5016)

Add a forProtocol classmethod to factory to easily build a Factory instance for
a protocol class.
